contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+
+### Examples
+* In multizone clusters, the PV controller has a hardcoded heuristic to provision
+PVCs for StatefulSets spread across zones.  If that zone does not have enough
+cpu/memory capacity to fit the pod, then the pod is stuck in pending state because
+its volume is bound to that zone.
+* Local storage exasperates this issue.  The chance of a node not having enough
+cpu/memory is higher than the chance of a zone not having enough cpu/memory.
+* Local storage PVC binding does not have any node spreading logic.  So local PV
+binding will very likely conflict with any pod anti-affinity policies if there is
+more than one local PV on a node.
+* A pod may need multiple PVCs.  As an example, one PVC can point to a local SSD for
+fast data access, and another PVC can point to a local HDD for logging.  Since PVC
+binding happens without considering if multiple PVCs are related, it is very likely
+for the two PVCs to be bound to local disks on different nodes, making the pod
+unschedulable.
+* For multizone clusters and deployments requesting multiple dynamically provisioned
+zonal PVs, each PVC Is provisioned independently, and is likely to provision each PV
+In different zones, making the pod unschedulable.
+
+To solve the issue of initial volume binding and provisioning causing an impossible
+pod placement, volume binding and provisioning should be more tightly coupled with
+pod scheduling.
+
+
+## Background
+In 1.7, we added alpha support for local PVs (TODO: link) with node affinity.
+You can specify a PV object with node affinity, and if a pod is using such a PV,
+the scheduler will evaluate the PV node affinity in addition to the other
+scheduling predicates.  So far, the PV node affinity only influences pod
+scheduling once the PVC is already bound.  The initial PVC binding decision was
+unchanged.  This proposal addresses the initial PVC binding decision.
+
+
+## Design
+The design can be broken up into a few areas:
+* User-facing API to invoke new behavior
+* Integrating PV binding with pod scheduling
+* Binding multiple PVCs as a single transaction
+* Recovery from kubelet rejection of pod
+* Making dynamic provisioning topology-aware
+
+For the alpha phase, only the user-facing API and PV binding and scheduler
+integration are necessary.  The remaining areas can be handled in beta and GA
+phases.
+
+### User-facing API
+This new binding behavior will apply to all unbound PVCs and can be controlled
+through install-time configuration passed to the scheduler and controller-manager.
+
+In alpha, this configuration can be controlled by a feature gate,
+`VolumeTopologyScheduling`.
+
+#### Alternative
+An alternative approach is to only trigger the new behavior only for volumes that
+have topology constraints.  A new annotation can be added to the StorageClass to
+indicate that its volumes have constraints and will need to use the new
+topology-aware scheduling logic.
```

Comment:
That is to say: when pv controller synchronized unbound pvc, it will always have to check storageclass's annotation to see which volume binding path it should go?

Reply:
Yes the alternative is to have the trigger be in the storageclass, so you need to check for the annotation every time you handle unbound volumes to determine if you should delay binding or not.